Indices Transforms

isaaccorley opened the pull request:
~~This is simply a draft of a transform for computing NDVI and concatenating to the `sample["image"]` channels. Open to suggestions/feedback.~~

~~Note: I just realized that the requiring the index of the channels for computing the indices doesn't scale. See [Enhanced Vegation Index](https://www.usgs.gov/core-science-systems/nli/landsat/landsat-enhanced-vegetation-index?qt-science_support_page_related_con=0#qt-science_support_page_related_con) which requires 4+ bands to compute `EVI = G * ((NIR - R) / (NIR + C1 * R – C2 * B + L))`.~~

Update:

This PR adds the following:
- `AugmentationSequential` wrapper around `kornia.augmentation.AugmentationSequential` which supports our sample/batch dicts.
- `ndbi`, `ndsi`, `ndvi`, `ndwi` functionals which given specific multispectral band input will compute the associated index.
`AppendNDBI`, `AppendNDSI`, `AppendNDVI`, `AppendNDWI` transform modules which take as input a batch dict, computes the desired index, and appends to the channel dimension.
- Unit tests
- `kornia>=0.5.4` as dependency which is the version when `kornia.augmentation.AugmentationSequential` was added. Additionally, kornia's only dependency is pytorch.

TODO:
- Add jupyter notebook to `docs/tutorials` display the usage of indices transforms
- Fix notebook sphinx errors


Closes #112

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,65 @@
+Module.__module__ = "torch.nn"
+
+
+def ndvi(red: Tensor, nir: Tensor) -> Tensor:
```
One of the things torchvision does that I've never understood is to have separate "functional" transforms in addition to regular transforms. Do we need these to be separate methods, or can we just have a single method?

isaaccorley replied:
I believe the only reasoning behind having a functional transform is if you do not know some parameters (e.g. dynamic input shapes) until performing a forward pass. In this case though we can remove the function.

isaaccorley replied:
I changed the class version of the indices transforms to `AppendNDVI`, for example. The reasoning being that `ndvi` performs the actual computation, while `AppendNDVI` is a transform which facilitates taking an input dict, computing the desired index, and appending it to the channel dimension. Let me know if this works if what naming scheme you would suggest.

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,65 @@
+    return (nir - red) / (nir + red)
+
+
+class NDVI(Module):  # type: ignore[misc,name-defined]
```
According to the [source code](https://github.com/kornia/kornia/blob/master/kornia/augmentation/container/augment.py#L87), our transforms need to subclass `_AugmentationBase` in order to be used with `AugmentationSequential`.

isaaccorley replied:
True, however computing an index and appending it as an additional channel/band is probably not considered an `augmentation` according to the `_AugmentationBase` class which requires probability of occurrence parameters like `p`, and `p_batch`. Although we could force these values to be `1.0` by default, I think that it's possible that `AugmentationSequential` is flexible enough to simply take in `nn.Module` transforms as well.

adamjstewart replied:
> I think that it's possible that AugmentationSequential is flexible enough to simply take in nn.Module transforms as well.

It might run, but mypy won't allow it.

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,65 @@
+
+    If you use this dataset in your research, please cite the following paper:
+
+    * https://arxiv.org/abs/1807.05713
```
Why this specific paper? NDVI has been around for decades and this particular paper is brand new.

isaaccorley replied:
Oops, this is actually the wrong paper.

isaaccorley replied:
Updated to correct paper.

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,65 @@
+        self.index_red = index_red
+        self.index_nir = index_nir
+
+    def forward(self, sample: Dict[str, Tensor]) -> Dict[str, Tensor]:
```
Do the other Kornia transforms take in a dict or a Tensor?

isaaccorley replied:
`AugmentationSequential` unfortunately takes in `n` tensors as the args in the `forward` method and uses the `data_keys` to identify them. We will likely need a wrapper to make it work with out dict. IMO it would make more sense to accept `data_keys` as a parameters and then take as input a dict and only modify the `data_keys`. Perhaps I should make an issue on the kornia repo about this.

Example:

```python
import torch
import kornia.augmentation as K

transforms = K.AugmentationSequential(
  K.RandomResizedCrop(size=(256, 256)),
  K.ColorJitter(),
  data_keys=["input", "mask"]
)
batch = dict(
 image=torch.randn(2, 3, 256, 256),
 mask=torch.zeros(2, 256, 256)
)
batch["image"], batch["mask"] = transforms(batch["image"], batch["mask"])
```

adamjstewart reviewed (changes requested):
This might need a bit more thought before we can safely integrate this with Kornia.

isaaccorley commented:
I'm curious if we should even utilize kornia to facilitate augmentations when these are instead simply transforms. We could use it along with kornia and the AugmentationSequential wrapper I created like below:

```python
import torch
import torch.nn as nn
import kornia.augmentation as K
from torchgeo.transforms import AugmentationSequential, AppendNDVI, AppendNDBI

sample = dict(
   image=torch.randn(2, 9, 512, 512).to("cuda"),
   mask=torch.randint(0, 10, size=(2, 1, 512, 512)).to(torch.float).to("cuda") # kornia requires mask to be float for some reason
)
transforms = nn.Sequential(
   AppendNDVI(index_red=2, index_nir=3),
   AppendNDBI(index_swir=4, index_nir=3),
   AugmentationSequential(
      K.RandomHorizontalFlip(),
      K.RandomVerticalFlip(),
      data_keys=["image", "mask"]
   )   
)
transforms = transforms.to("cuda")
outputs = transforms(sample)
# "image": (2, 11, 512, 512)
# "mask": (2, 1, 512, 512)
```

The above actually works as of now and simply chains multiple non augmentation transforms (e.g. computing indices), along with  an AugmentationSequential wrapper which then performs the transforms on the image and mask.

calebrob6 commented:
The above seems fine to me! Thoughts:
- This seems remote sensing specific and doesn't need to be added to kornia
- Things we write here should be explicit whether they work on batches or samples (or both!). From the limited amount I've seen, the kornia augmentation pipelines operate on batches, while torchvision transforms operate on samples.

isaaccorley commented:
That's a great point about sample vs batch. Apparently torchvision transforms on tensors work for both batch and sample data. Although, under the hood, I believe they just add a batch dim of 1 to sample data and then squeeze it off afterwards.

A note on the AugmentationSequential wrapper. I'm actually hoping that a PR I'm working on for adding this to kornia will be accepted. Until then, using the wrapper should allow us to chain transforms and augs similarly to the above example.

isaaccorley commented:
I'm running into a very strange mypy error due to something in kornia. Tracking it down...
```
mypy .
Traceback (most recent call last):
  File "/home/icorley/miniconda3/envs/torchgeo/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/home/icorley/miniconda3/envs/torchgeo/lib/python3.9/site-packages/mypy/__main__.py", line 11, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "mypy/main.py", line 87, in main
  File "mypy/main.py", line 165, in run_build
  File "mypy/build.py", line 179, in build
  File "mypy/build.py", line 254, in _build
  File "mypy/build.py", line 2697, in dispatch
  File "mypy/build.py", line 3021, in process_graph
  File "mypy/build.py", line 3138, in process_stale_scc
  File "mypy/build.py", line 2288, in write_cache
  File "mypy/build.py", line 1475, in write_cache
  File "mypy/nodes.py", line 313, in serialize
  File "mypy/nodes.py", line 3149, in serialize
  File "mypy/nodes.py", line 3083, in serialize
AssertionError: Definition of kornia.geometry.subpix.spatial_soft_argmax.normalize_pixel_coordinates is unexpectedly incomplete
```

calebrob6 commented:
> I'm running into a very strange mypy error due to something in kornia. Tracking it down...

I'm getting this same error in the trainers PR

calebrob6 commented:
spatial_soft_argmax.py needs to do `from kornia.geometry.conversions import normalize_pixel_coordinates, normalize_pixel_coordinates3d` instead of `from kornia.geometry...`

isaaccorley commented:
Ok, I'll make an issue in the kornia repo.

isaaccorley commented:
Sphinx seems to be failing due to warning that it doesn't  allow cells with shell commands in them (e.g. `!pip install kornia`). I could rewrite using Python but I'm curious if we even want this level of strictness?

adamjstewart commented:
> Sphinx seems to be failing due to warning that it doesn't allow cells with shell commands in them (e.g. `!pip install kornia`). I could rewrite using Python but I'm curious if we even want this level of strictness?

That's odd, can you open an issue at https://github.com/spatialaudio/nbsphinx/issues? We definitely want/need this to be supported. This is a very common thing to have in a notebook and is a requirement for our code to work on Colab.

P.S. We need to convert our other notebooks to `pip install torchgeo...` at some point. Since our GitHub Actions for testing notebooks already runs `pip install` we don't have to worry about it installing a different version from PyPI. I don't love that it can't be tested locally though. Haven't figured that one out yet.

adamjstewart commented on setup.cfg:
```diff
@@ -30,6 +30,8 @@ install_requires =
     einops
     # fiona 1.5+ required for fiona.transform module
     fiona>=1.5
+    # kornia latest required for mypy fixes
```
We only ever use the latest version when testing mypy anyway, mypy issues aren't reasons to make a version required, I would go back to the previous requirements in setup.cfg and environment.yml but install a newer version just for CI.

isaaccorley replied:
Reverted and added to the tests github action under the mypy run section.

adamjstewart commented on .github/workflows/tests.yaml:
```diff
@@ -21,6 +21,7 @@ jobs:
       run: |
         pip install cython numpy  # needed for pycocotools
         pip install .[datasets,tests,train]
+        pip install git+https://github.com/kornia/kornia@master # required for mypy fixes
```
This might need to be first to prevent installing twice

isaaccorley replied:
Done

adamjstewart commented on setup.cfg:
```diff
@@ -48,7 +50,6 @@ install_requires =
     torchvision>=0.3
 python_requires = >= 3.6
 packages = find:
-
```
Can we keep this empty line?

isaaccorley replied:
Fixed

adamjstewart commented on .github/workflows/tests.yaml:
```diff
@@ -19,6 +19,7 @@ jobs:
         python-version: 3.9
     - name: Install pip dependencies
       run: |
+        pip install git+https://github.com/kornia/kornia@master # required for mypy fixes
```
Can you add a TODO to remove this once the next Kornia release comes out?

isaaccorley replied:
Added TODO

adamjstewart commented on docs/index.rst:
```diff
@@ -28,6 +28,7 @@ torchgeo
    tutorials/getting_started
    tutorials/trainers
    tutorials/benchmarking
+   tutorials/indices
```
I haven't looked through the tutorial notebook yet but I would like it to be a generic "here is how to use both Kornia and torchgeo transforms separately and together" instead of just a tutorial on indices. So the file should be named `transforms.ipynb`.

isaaccorley replied:
The tutorial shows how to use the functional indices (e.g. (ndvi), the indices modules (e.g. AppendNDVI), and chaining the indices modules together using the AugmentationSequentialWrapper and other kornia augmentations and applying them to a batch dict of imagery.

The reason I sertled on the indices title is because I also added an additional section on how to extract and download a smaller region of interest from Sentinel-2 COGs and then us our functional indices to compute indices and then visualize them.

It might make more sense to split them into two separate tutorials though.

isaaccorley replied:
I split them into separate notebooks, indices.ipynb and transforms.ipynb.

adamjstewart replied:
I don't think we need a notebook for each type of transform, what do you think @calebrob6?

isaaccorley replied:
I was thinking of indices as their own thing since, if you look at the notebook, it's centered around using them to create index visualizations on raw Sentinel-2 imagery. The transforms notebook I was thinking would just be for all other general transforms.

adamjstewart commented on setup.cfg:
```diff
@@ -30,6 +30,8 @@ install_requires =
     # fiona 1.5+ required for fiona.transform module
     fiona>=1.5
+    # kornia 0.5.4+ required for kornia.augmentation.AugmentationSequential
+    kornia>=0.5.4
```
Is Kornia actually used in torchgeo or just in the tests/tutorials?

isaaccorley replied:
It's used in the AugmentationSequential wrapper in transforms.py but otherwise only in tests, trainers, and tutorials.

adamjstewart commented on tests/transforms/test_transforms.py:
```diff
@@ -68,3 +84,43 @@ def test_identity(sample: Dict[str, Tensor]) -> None:
+    assert_matching(output, expected)
+
+
+def test_augmentation_sequential_image_only(batch: Dict[str, Tensor]) -> None:
```
We need tests that use this for:

1. Only torchgeo transforms
2. Only Kornia transforms
3. Torchgeo + Kornia transforms

isaaccorley replied:
I added more tests for the above as well as testing using a grayscale, rgb, and multispectral images. Let me know if you want any more.

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,234 @@
+Module.__module__ = "torch.nn"
+
+
+epsilon = 1e-10
```
```suggestion
_EPSILON = 1e-10
```
Should be a private global variable

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,234 @@
+                nir=sample["image"][:, self.index_nir],
+            )
+            index = index.unsqueeze(self.dim)
+            sample["image"] = torch.cat([sample["image"], index], dim=self.dim)  # type: ignore[attr-defined]  # noqa: E501
```
If you split this into multiple lines you won't need to ignore line length here. Like:
```python
sample["image"] = torch.cat(  # type: ignore[attr-defined]
    ...
)
```

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,234 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""TorchGeo indices transforms."""
```
It might be useful to add links to those resources you used for index definitions and citations so that we don't forget about them.

isaaccorley replied:
Added links, let me know if it works for you.

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,234 @@
+        Args:
+            index_green: index of the Green band in the image
+            index_swir: index of the Short-wave Infrared (SWIR) band in the image
+            dim: dimension of channels in the input tensors (default: 1)
```
What is this used for, the channel dimension? Could we just always use -3 instead of requiring the user to specify it?

isaaccorley replied:
We definitely can. In the future we may want to add support for computing indices over time series imagery (b,t,c,h,w) but I think we can just take the shapes and unsqueeze them to agree so that they can be concatenated instead of requiring a user specified dim.

isaaccorley replied:
Removed dim argument and hardcoded to `self.dim = -3`

adamjstewart commented on torchgeo/transforms/indices.py:
```diff
@@ -0,0 +1,234 @@
+                nir=sample["image"][:, self.index_nir],
+            )
+            index = index.unsqueeze(self.dim)
+            sample["image"] = torch.cat([sample["image"], index], dim=self.dim)  # type: ignore[attr-defined]  # noqa: E501
```
Same here, this should be wrapped to multiple lines

adamjstewart commented on torchgeo/transforms/transforms.py:
```diff
@@ -97,3 +98,57 @@ def forward(self, sample: Dict[str, Tensor]) -> Dict[str, Tensor]:
+            if key == "image":
+                keys.append("input")
+            elif key == "boxes":
+                keys.append("bbox")
```
This line isn't being tested

isaaccorley replied:
Added tests with bounding boxes.

adamjstewart commented on torchgeo/transforms/transforms.py:
```diff
@@ -97,3 +98,57 @@ def forward(self, sample: Dict[str, Tensor]) -> Dict[str, Tensor]:
+        }
+        sample.update(outputs)
+
+        # Convert masks to int
```
Are we guaranteed that all masks will be integers? Is there a reason we need to do this?

isaaccorley replied:
That's a good point. I don't believe I've seen a case where you would want masks to not be ints since Cross Entropy losses generally require the targets to be ints.

Possibly student/teacher distillation type experiments where the targets would be the soft probabilities? Even then the masks would only be computed in the forward pass and not provided by the dataloader.

Anyways, maybe a solution is to check the dtype of the mask prior to augmentation and just recast to that dtype post augmentation.

isaaccorley replied:
Updated to cast masks after augmentation to whatever dtype they were prior to augmentation.

adamjstewart commented on torchgeo/transforms/transforms.py:
```diff
@@ -97,3 +98,57 @@ def forward(self, sample: Dict[str, Tensor]) -> Dict[str, Tensor]:
+        Returns:
+            the augmented input
+        """
+        # Kornia augmentations require masks to be float
```
Why is that?

isaaccorley replied:
@calebrob6 ran into this as well. For the ops that apply to both images and masks the augmentations raise an error that the inputs should be some type of float. I simply added this to avoid a user having to do this every time they want to use AugmentationSequential. This is a temporary fix though as I'm working on a kornia PR where kornia.AugmentationSequential does the casting under the hood.

adamjstewart reviewed (changes requested):
Should we get rid of `torchgeo/transforms/transforms.py`? We're already planning on dropping the random flip transforms, and we could move `Identity` to the tests themselves. I'm not sure where the best place to put `AugmentationSequential` would be. It could be in `__init__.py` itself or in `torchgeo/transforms/containers.py`.

I still need to look at the tutorial. I still have a lot of questions remaining about how Kornia does things and how TorchGeo is compatible with Kornia, but I'll save those for later.

isaaccorley commented:
Great question, if kornia adds support for optionally taking a dict input and returning a dict output to AugmentationSequential where only the specified keys are modified, we could actually drop the AugmentationSequential wrapper entirely and transforms.py as well. Then kornia would only be used in our trainers and tutorials.

isaaccorley commented:
> > Sphinx seems to be failing due to warning that it doesn't allow cells with shell commands in them (e.g. `!pip install kornia`). I could rewrite using Python but I'm curious if we even want this level of strictness?
> 
> That's odd, can you open an issue at https://github.com/spatialaudio/nbsphinx/issues? We definitely want/need this to be supported. This is a very common thing to have in a notebook and is a requirement for our code to work on Colab.
> 
> P.S. We need to convert our other notebooks to `pip install torchgeo...` at some point. Since our GitHub Actions for testing notebooks already runs `pip install` we don't have to worry about it installing a different version from PyPI. I don't love that it can't be tested locally though. Haven't figured that one out yet.

Just to circle back to this. I was able to get around this by using cell magic `%` instead of exclamation marks `!` for executing things like `%pip install` or directory manipulation (e.g. `cd`, `ls`, etc.).

isaaccorley commented:
Kornia's `AugmentationSequential` seems to be broken for bounding boxes. Even when using only `AugmentationSequential(K.RandomHorizontalFlip(p=1.0, p_batch=1.0), data_keys=["image", "mask", "boxes"])` I get different values each time I feed our dicts through. Sometimes I get negative transformed bbox values (???).

I'll make an issue on their repo.

adamjstewart commented:
> Something to point out is that we don't actually need our version of torch.nn.Identity other than to be compatible with type checking since nn.Identity input/output is a Tensor and ours is a dict. Both simply return the input as output.

Didn't realize this already existed, perfect!

adamjstewart commented:
Made the following changes to the notebooks:

- [x] Moved the tutorials higher up in the list since they will be more commonly used
- [x] Removed the "Open in Colab" buttons, these are added by nbsphinx at document generation time
- [x] Switched the pip install to the main repo
- [x] Download dataset to a temp directory
- [x] Converted to 4-space indentation
- [x] Replaced `torch.clip` with `torch.clamp`, see #160

Remaining TODOs which can happen in a follow-up PR:

- [ ] Need to switch to the new EuroSat dataset, or use a different dataset with higher res imagery
- [ ] There isn't a lot of "why we are doing this" in the Transforms tutorial. For example, why are we implementing our own MinMaxNormalize? Is there none in Kornia? Is this just for demonstration purposes so users know how to write their own transforms?
- [ ] What's all this stuff?
    ```
    #@title EuroSat Multispectral (MS) Browser  { run: "auto", vertical-output: true }
    idx = 9889 #@param {type:"slider", min:0, max:26999, step:1}
    ```
- [ ] The Sentinel image we plot in the Transform tutorial is very hard to see, I would plot the raw image instead of the 0-1 scaled image, see the `plot` method in `GeoDataset`
- [ ] Transform tutorial needs more visualization
- [ ] Indices transform might be a bit too complicated
- [ ] For the indices, you say they can be calculated from the formula below, but there is no formula

The Indices tutorial has a suspiciously greater amount of comments than the Transforms tutorial. You didn't get the Indices tutorial from somewhere else did you?

isaaccorley commented:
I added additional comments for clarity, so no, I didn't pull this tutorial from somewhere else.

I would recommend opening the notebook in colab otherwise you can't see the visualization app that uses some of the Colab functionality. The image that's plotted might be dark because it's one of the images of a lake/sea.

I'll refactor to use kornia's minmax normalize. Although, they don't have a stretch/percentile normalize so I'll have to hack something together.

isaaccorley commented:
I'll make some of the changes in a but and the rest can come in a separate PR I think.

isaaccorley commented:
Some updates:

- [X] Need to switch to the new EuroSat dataset, or use a different dataset with higher res imagery
- Switched to `torchgeo.datasets.EuroSAT` and which removed a lot of boilerplate code.
- [X] There isn't a lot of "why we are doing this" in the Transforms tutorial. For example, why are we implementing our own MinMaxNormalize? Is there none in Kornia? Is this just for demonstration purposes so users know how to write their own transforms?
- I added some more markdown sections to shed light on why. Also added a quick benchmark comparison of using the transforms on cpu/gpu and timed them to show why you should prefer chained transforms in a `nn.Sequential` that can be pushed to the GPU. There is a [`kornia.enhance.normalize_min_max`](https://github.com/kornia/kornia/blob/49440520494fe799d1fbbccddf0cbc453cdbced4/kornia/enhance/normalize.py#L243) but there is no class form of it, and it only takes in a single min/max val instead of a list to min/max normalize per channel. Will make a PR in kornia's repo for this.
- [X] What's all this stuff?
```
#@title EuroSat Multispectral (MS) Browser  { run: "auto", vertical-output: true }
idx = 9889 #@param {type:"slider", min:0, max:26999, step:1}
```
- Added a comment to relay that this can only be viewed in Colab
- [X] The Sentinel image we plot in the Transform tutorial is very hard to see, I would plot the raw image instead of the 0-1 scaled image, see the plot method in GeoDataset
- Changed the default displayed image so it's easier to see. It's also blurry because they are 64x64 images that when displayed are very small so I am interpolating them for better visualization.
- [X] Transform tutorial needs more visualization
- Will add this in a separate PR
- [X] Indices transform might be a bit too complicated
- I could probably remove the COG windowed download parts but thought that might be helpful.
- [X] For the indices, you say they can be calculated from the formula below, but there is no formula
- Added the formulas

adamjstewart commented:
> Added a comment to relay that this can only be viewed in Colab

The primary purpose of the tutorials is to serve as documentation, the fact that they can be ran on Colab is less important. If this shows up on Read the Docs and doesn't do anything outside of Colab I think we should just remove it. Aren't there Jupyter Notebook magic things to make this work everywhere?

adamjstewart reviewed (approved):
I'm going to merge this as is. At some point before we release I plan on going back through the tutorials and touching them up and testing them, so if there's anything else needed I can work on them then.
